sdk/eventhub/azure-eventhub/azure/eventhub/_buffered_producer/_buffered_producer_dispatcher.py

```diff
@@ -0,0 +1,147 @@
+        self._existing_executor = bool(executor)
+        self._executor = executor or ThreadPoolExecutor(max_worker)
+
+    def _get_partition_id(self, partition_id, partition_key):
+        if partition_id:
+            if partition_id not in self._partition_ids:
+                raise ConnectError(
+                    "Invalid partition {} for the event hub {}".format(
+                        partition_id, self._eventhub_name
+                    )
+                )
+            return partition_id
+        if isinstance(partition_key, str):
+            return self._partition_resolver.get_partition_id_by_partition_key(partition_key)
+        return self._partition_resolver.next_partition_id
+
+    def enqueue_events(self, events, *, partition_id=None, partition_key=None, timeout=None):
+        pid = self._get_partition_id(partition_id, partition_key)
+        with self._lock:
+            try:
+                self._buffered_producers[pid].put_events(events, timeout)
+            except KeyError:
+                buffered_producer = BufferedProducer(
+                    self._create_producer(pid),
+                    pid,
+                    self._on_success,
+                    self._on_error,
+                    self._max_message_size_on_link,
+                    executor=self._executor,
+                    max_wait_time=self._max_wait_time,
+                    max_concurrent_sends=self._max_concurrent_sends,
+                    max_buffer_length=self._max_buffer_length
+                )
+                buffered_producer.start()
+                self._buffered_producers[pid] = buffered_producer
+                buffered_producer.put_events(events, timeout)
+
+    def flush(self, timeout=None):
+        # flush all the buffered producer, the method will block until finishes or times out
+        with self._lock:
+            futures = []
+            for pid, producer in self._buffered_producers.items():
+                # call each producer's flush method
+                futures.append((pid, self._executor.submit(producer.flush, timeout=timeout)))
+
+            # gather results
+            exc_results = {}
+            for pid, future in futures:
+                try:
+                    future.result()
+                except Exception as exc:
+                    exc_results[pid] = exc
+
+            if not exc_results:
+                _LOGGER.info("Flushing all partitions succeeded")
+                return
+
+            _LOGGER.info('Flushing all partitions partially succeeded with result {!r}.'.format(exc_results))
+            # TODO: better error?
+            raise EventDataSendError(
+                message="Flushing all partitions partially succeeded, failed partitions are {!r}.".format(
+                    exc_results.keys()
+                ),
+                details=exc_results
+            )
+
+    def close(self, flush=True, timeout=None, raise_error=False):
```

Comment:
Nit: Even though it's internal, I feel keyword-only syntax would be nice here....

Reply:
updated!